### PR TITLE
[DTensor] Preserve symbolic local layouts without DDE guards

### LINKED ISSUE #187025: [DTensor] to_local backward should preserve same-placement symbolic layout without recomputing stride
## Summary

Compiled DTensor `to_local()` backward can fail with unbacked symbolic shapes
when the backward path keeps the same DTensor placements as the forward input.

`_ToTorchTensor.backward` currently recomputes global gradient stride from the
local gradient tensor before deciding whether it can reuse the original DTensor
spec. Under AOTAutograd tracing, the local gradient stride can be symbolically
equivalent to the original local stride but carry different symbol provenance.
For example, the saved DTensor spec can contain `(u1, 1)` while the local
gradient stride is `(Max(1, u3), 1)`.

The recomputation goes through `compute_global_tensor_info()` and can force a
data-dependent guard before the code has a chance to decide that same-placement
backward should preserve the original DTensor layout.

## Reproducer

```python
import torch
import torch.distributed as dist
from torch.distributed import DeviceMesh
from torch.distributed.tensor import DTensor, Shard
from torch.testing._internal.distributed.fake_pg import FakeStore


try:
    dist.destroy_process_group()
except Exception:
    pass

dist.init_process_group("fake", store=FakeStore(), rank=0, world_size=2)
mesh = DeviceMesh("cpu", torch.arange(2))

local = torch.randn(8, 8, requires_grad=True)
dt = DTensor.from_local(local, mesh, [Shard(0)], run_check=False)
torch._dynamo.decorators.mark_unbacked(dt, 0, hint_override=dt.shape[0])
torch._dynamo.decorators.mark_unbacked(dt, 1, hint_override=dt.shape[1])


@torch.compile(backend="aot_eager", fullgraph=True)
def fn(x):
    return (x.to_local() * 2).sum()


fn(dt).backward()
print(local.grad)
```

## Actual behavior

Compilation fails during `_ToTorchTensor.backward`:

```text
GuardOnDataDependentSymNode: Could not guard on data-dependent expression Max(1, u3) <= 1
Caused by: (distributed/tensor/_api.py in backward)
```

## Expected behavior

The compiled backward should complete and produce the same local gradient as the
uncompiled run.

For same-placement `to_local()` backward, DTensor should preserve the forward
DTensor layout unless the local gradient stride is provably different. Symbolic
strides that may match but have different provenance should not force global
stride recomputation or data-dependent guards.

cc @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk @weifengpy @chauhang @penguinwu @ezyang @bobrenjc93 @laithsakka @tianyu-l @XilunWu @SherlockNoMad @ppwwyyxx @bdhirsh @aorenste

### PR DESCRIPTION
**Pull request by sanketpurandare**
Compiled DTensor paths can see symbolic local layout metadata that is
semantically valid but not syntactically identical to the metadata saved during
forward propagation.

For `to_local()` backward, AOTAutograd can produce a local gradient stride like
`(Max(1, u3), 1)` while the saved DTensor metadata uses `(u1, 1)`. The previous
backward path recomputed the global gradient stride before deciding whether it
could reuse the original DTensor spec. That forced
`compute_global_tensor_info()` to evaluate symbolic stride relations and could
raise a data-dependent guard for the default same-placement backward path.

Reuse the original DTensor spec only for the default same-placement backward
when the local gradient stride, saved forward local stride, and DTensor spec
stride are compatible. Exact/provable stride equality is accepted directly. For
contiguous symbolic stride forms such as `Max(1, u*)`, use the existing
`check_contiguous_sizes_strides(..., false_if_dde=True)` helper so equivalent
contiguous layouts are recognized without requiring a brittle exact symbolic
match. If neither exact nor contiguous equivalence can be proven, emit
`torch._check` assertions for the required stride equalities before taking the
symbolic shortcut.

If the placement changes or the physical local stride cannot justify the
original spec layout, keep the existing recomputation path and build a fresh
spec from the observed gradient stride. This avoids the symbolic guard failure
without lying about memory layout. In particular, uneven channels-last shards
must keep using the recomputation path so autograd can repair the physical local
gradient layout correctly.

The same class of issue also appears in `aten.t` sharding propagation. The
single-dim strategy can enumerate candidate placements that move a symbolic
`_StridedShard` split factor onto the other tensor dimension. When that
candidate is not provably shardable, strategy expansion should reject it instead
of evaluating a Python bool on an unbacked expression such as `8 < 2*u0`.
Register transpose with `allow_unbacked_sharding=False` so unproven candidates
are pruned while statically valid candidates, including `_StridedShard(0, u0)`
propagating to `_StridedShard(1, u0)`, are still kept.

Fixes #187025

This PR was authored with assistance from an AI assistant.

Test Plan:

```bash
python -m pytest test/distributed/tensor/test_op_strategy.py::TestCostModel::test_t_prunes_unproven_unbacked_strided_shard_candidates test/distributed/tensor/test_op_strategy.py::TestCostModel::test_mm_strategies test/distributed/tensor/test_dtensor_compile.py::TestDTensorCompile::test_to_local_backward_unbacked_symbolic_stride test/distributed/tensor/test_tensor_ops.py::TestNewEmptyStridedUneven::test_backward_channels_last -q -s
```

```bash
lintrunner -a
```

### CODE REVIEW

**Review by pianpwk — APPROVED** [review]
Maybe you need something like https://github.com/pytorch/pytorch/blob/ca8348e4dd0d25cdf6e7336520cbe8f57072c893/torch/_prims_common/__init__.py#L259-301 ? 

Not sure if this case is contiguous, but `Max(u*, 1)` expressions show up w/ `x.contiguous().stride()`, though they might get permuted after.

Maybe something like 
```
if (
    grad_placements == dtensor_spec.placements
    and (
         torch._prims_common.same_shape(tensor_stride, dtensor_meta.stride)
         or check_contiguous_sizes_strides(tensor_stride, dtensor_meta.stride)
    )
):
```

is sufficient? maybe that's a stretch.

Stamping to unblock. cc @laithsakka

**Comment by sanketpurandare** [comment]
> Maybe you need something like https://github.com/pytorch/pytorch/blob/ca8348e4dd0d25cdf6e7336520cbe8f57072c893/torch/_prims_common/__init__.py#L259-301 ?
> 
> Not sure if this case is contiguous, but `Max(u*, 1)` expressions show up w/ `x.contiguous().stride()`, though they might get permuted after.
> 
> Maybe something like
> 
> ```
> if (
>     grad_placements == dtensor_spec.placements
>     and (
>          torch._prims_common.same_shape(tensor_stride, dtensor_meta.stride)
>          or check_contiguous_sizes_strides(tensor_stride, dtensor_meta.stride)
>     )
> ):
> ```
> 
> is sufficient? maybe that's a stretch.
> 
> Stamping to unblock. cc @laithsakka

Great suggestion, made the change
